bench/macro/lsm-tree-bench-unions.hs

```diff
@@ -0,0 +1,629 @@
+
+    -- Setup RNG for randomized lookups
+    initializeRNG gopts
+
+    -- Perform 3 measurement phases
+    --   * Phase 1: Measure performance before supplying any credits.
+    --   * Phase 2: Measure performance as credits are incrementally supplied and debt is repaid.
+    --   * Phase 3: Measure performance when debt is 0.
+    let tickCountPrefix = 50
+        tickCountMiddle = 100
+        tickCountSuffix = 50
+        tickCountEnding = maximum indicesPhase3
+        samplingLookups = batchSizeOfSample opts
+        samplesEachTick = batchSamplesPerTick opts
+        queriesEachTick = samplesEachTick * samplingLookups
+        indicesPhase1 = [ negate tickCountPrefix .. 0 ]
+        indicesPhase2 = [ 1 .. tickCountMiddle ]
+        indicesPhase3 = [ tickCountMiddle + 1 .. tickCountMiddle + tickCountSuffix ]
+        indicesDomain = indicesPhase1 <> indicesPhase2 <> indicesPhase3
+
+    measurements <- LSM.withSession (rootDir gopts) $ \session -> do
+        -- Load the baseline table
+        table_0 <- LSM.openTableFromSnapshot session baselineTableName label
+
+        -- Load the union tables
+        tables <- forM (tableRange gopts) $ \tID ->
+            LSM.openTableFromSnapshot session (makeTableName tID) label
+
+        LSM.withIncrementalUnions tables $ \table -> do
+          LSM.UnionDebt totalDebt <- LSM.remainingUnionDebt table
+          -- Determine the number of credits to supply per tick in order to
+          -- all debt repaid at the time specified by the rpayment rate.
+          -- Each tick should supply credits equal to:
+          --     paymentRate * totalDebt / tickCountMiddle
+          let paymentPerTick = ceiling $ toInteger totalDebt % tickCountMiddle
+
+          let measurePerformance :: Integer -> IO (Int, Int, TimingResult, TimingResult)
+              measurePerformance tickIndex = do
+                -- Note this tick's debt for subsequent measurement purposes.
+                LSM.UnionDebt debtCurr <- LSM.remainingUnionDebt table
+                -- Note the cumulative credits supplied through this tick.
+                let paidCurr = max 0 $ totalDebt - fromInteger (max 0 tickIndex) * paymentPerTick
+                -- Generate the randomized lookup batches.
+                -- There are two important aspects to consider when preparing for the measurements.
+                --   1. Generate all the lookup keys *before* starting the measurement!
+                --      This ensures that the benchmark does not include RNG latency in the results.
+                --   2. Use these exact same lookups for both the baseline table and the unioned table.
+                --      This ensures that the comparison between a monolithic table and union of tables
+                --      are appropriately comparable.
+                sampleSeriesOfBatches <- fmap V.force . V.replicateM samplesEachTick $
+                    generateBatchSampling (initialSize gopts) samplingLookups
```

Comment:
I've mentioned it elsewhere too, but I think it's okay generate lookup batches on the fly, instead of pre-generating them

Reply:
I replace the pre-comuted array with batches generated "on the fly." I supplied cloned copies of the same RNG to each benchmark measurement for the "baseline" and "unioned" tables to ensure that the same batches of keys are generated in each measurement. This ought to ensure that the measurements are comparable.